### PR TITLE
feat(model): add upsertKeys option to bulkCreate and upsert

### LINKED ISSUE #13605: Support custom upsertKeys for bulkCreate
## Issue Creation Checklist

- [X] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

### Is your feature request related to a problem? Please describe.

I am trying to perform a bulk update of rows where I already know the PK, and each row needs to be updated with a different value. The only way to do this is with a `bulkCreate` while specifying `updateOnDuplicate`. One of the fields I would like to update is part of the set of unique keys, so the default functionality of using the unique keys fails for this case and returns a unique key error (tries to insert instead).

### Describe the solution you'd like

The `bulkCreate` (and upsert by extension) should accept a set of keys to use as the uniqueness constraint for the `updateOnDuplicate` logic.

```js
const people = await Person.bulkCreate(updates, {
  updateOnDuplicate: ['emailAddress', 'name'],
  upsertKeys: ['emailAddress']
});
```

### Why should this be in Sequelize

This adds the ability to configure a core aspect of an `updateOnDuplicate` query, and therefore should be supported by the raw query generator

### Describe alternatives/workarounds you've considered

- manually passing in `upsertKeys` to query call without this PR (overwritten internally)
- changing the `upsertKeys` field on the options to a custom getter/setter to prevent from being overwritten (still overwritten internally)

### Is this feature dialect-specific?

- [ ] No. This feature is relevant to Sequelize as a whole.
- [X] Yes. This feature only applies to the following dialect(s): Same dialects as `updateOnDuplicate`; MySQL, MariaDB, SQLite >= 3.24.0 & Postgres >= 9.5

### Would you be willing to implement this feature by submitting a Pull Request?

- [X] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I don't know how to start, I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

### PR DESCRIPTION
**Pull request by blake-transcend**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [X] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [X] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [X] Have you added new tests to prevent regressions?
- [X] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [X] Did you update the typescript typings accordingly (if applicable)?
- [X] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description of change

- closes #13605
- Adds support to `bulkCreate` and `upsert` to accept user-specified `upsertKeys`

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -2533,6 +2533,7 @@ class Model {
    * @param  {boolean}        [options.benchmark=false]        Pass query execution time in milliseconds as second argument to logging function (options.logging).
    * @param  {boolean|Array}  [options.returning=false]        If true, append RETURNING <model columns> to get back all defined values; if an array of column names, append RETURNING <columns> to get back specific columns (Postgres only)
    * @param  {string}         [options.searchPath=DEFAULT]     An optional parameter to specify the schema search_path (Postgres only)
+   * @param  {Array}          [options.upsertKeys]             Fields to use as the unique keys (on duplicate key update)? (only supported by MySQL, MariaDB, SQLite >= 3.24.0 & Postgres >= 9.5). By default, unique keys and/or primary keys are used.
```
Comment (AllAwesome497) [inline lib/model.js:2536]: I don't see anything in the MySQL or MariaDB docs which show that this is actually supported.

They only have `ON DUPLICATE KEY` whereas for postgres and sqlite this is translated to `ON CONFLICT ({upsertKeys}) DO UPDATE ...`.

**Comment by AllAwesome497** [comment]
I already have a pull request for this, #13419.  I've also got 2 others to support `ON CONFLICT WHERE` conditions for both `Model.upsert` (#13419) and `Model.bulkCreate` (#13420).

I'll put up a pull request for specifying this in `Model.upsert` too, I just haven't taken the time to take this change from my fork and put it into a pull request.

**Comment by sdepold** [comment]
> I already have a pull request for this, #13419. I've also got 2 others to support `ON CONFLICT WHERE` conditions for both `Model.upsert` (#13419) and `Model.bulkCreate` (#13420).
> 
> I'll put up a pull request for specifying this in `Model.upsert` too, I just haven't taken the time to take this change from my fork and put it into a pull request.

Means we should rather close this one?

**Comment by AllAwesome497** [comment]
> Means we should rather close this one?

Yes, looks like it; closing.

**Pull request closed without merging**